src/vm/castcache.cpp

```diff
@@ -0,0 +1,291 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+//
+// File: castcache.cpp
+//
+
+#include "common.h"
+#include "castcache.h"
+
+#if !defined(DACCESS_COMPILE) && !defined(CROSSGEN_COMPILE)
+
+OBJECTHANDLE CastCache::s_cache = NULL;
+
+BASEARRAYREF CastCache::CreateCastCache(DWORD size)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    // size must be positive
+    _ASSERTE(size > 0);
+    // size must be a power of two
+    _ASSERTE((size & (size - 1)) == 0);
+
+    BASEARRAYREF table = NULL;
+
+    try
+    {
+        table = (BASEARRAYREF)AllocatePrimitiveArray(CorElementType::ELEMENT_TYPE_I4, (size + 1) * sizeof(CastCacheEntry) / sizeof(INT32));
+    }
+    catch (OutOfMemoryException)
+    {
+        // try a small cache
+        size = INITIAL_CACHE_SIZE;
+        try
+        {
+            table = (BASEARRAYREF)AllocatePrimitiveArray(CorElementType::ELEMENT_TYPE_I4, (size + 1) * sizeof(CastCacheEntry) / sizeof(INT32));
+        }
+        catch (OutOfMemoryException)
+        {
+            // OK, no cache then
+            return NULL;
+        }
+    }
+
+    TableMask(table) = size - 1;
+
+    // Fibonacci hash reduces the value into desired range by shifting right by the number of leading zeroes in 'size-1' 
+    DWORD bitCnt;
+#if BIT64
+    BitScanReverse64(&bitCnt, size - 1);
+    HashShift(table) = (BYTE)(63 - bitCnt);
+#else
+    BitScanReverse(&bitCnt, size - 1);
+    HashShift(table) = (BYTE)(31 - bitCnt);
+#endif
+
+    return table;
+}
+
+BOOL CastCache::MaybeReplaceCacheWithLarger(DWORD size)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    BASEARRAYREF newTable = CreateCastCache(size);
+    if (!newTable)
+    {
+        return FALSE;
+    }
+
+    OBJECTREF currentTableRef = ObjectFromHandle(s_cache);
+    OBJECTREF prevTableRef = (OBJECTREF)(Object*)InterlockedCompareExchangeObjectInHandle(s_cache, newTable, currentTableRef);
+
+    return prevTableRef == currentTableRef;
+}
+
+void CastCache::FlushCurrentCache()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    BASEARRAYREF currentTableRef = (BASEARRAYREF)ObjectFromHandle(s_cache);
+    int size = !currentTableRef ? INITIAL_CACHE_SIZE : CacheElementCount(currentTableRef);
+
+    BASEARRAYREF newTable = CreateCastCache(size);
+    StoreObjectInHandle(s_cache, newTable);
+}
+
+void CastCache::Initialize()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    s_cache = CreateGlobalHandle(NULL);
+}
+
+TypeHandle::CastResult CastCache::TryGet(TADDR source, TADDR target)
```

Comment:
Do we want to have this fast path forceinlined, in the most perf critical places at least?

Reply:
Looks a bit too big to me for inlining.  I think forceinline will more likely hurt than help. 
Also the most perf critical places are currently in asm. I think we should try forceinlining later.

---
In reply to: [335156444](https://github.com/dotnet/coreclr/pull/23548#discussion_r335156444) [](ancestors = 335156444)

Reply:
Still, it would be nice to look at the inlining. E.g. the path for `JIT_ChkCastArray` goes through (non-inlined) ObjIsInstanceOfCached and then goes to (non-inlined) CastCache::TryGet.

Reply:
I looked through most common paths. - Just to see if there is nothing obviously bad or easy opportunities.
`ObjIsInstanceOfCached` typically tail-calls TryGet, when is not inlined itself.

I tried forceinlining `ObjIsInstanceOfCached `, but that failed on Linux when called from asm. MSVC seems to handle that, just does not inline when it can't.
We could have two version - one to call from asm and another forceinlined - to call from C, but I figured it might be ok as-is for now. Not sure how many of those asm helpers will stay around eventually.

I looked a lot at `JIT_ChkCastAny`, which is basically identical to `JIT_ChkCastArray` now.
It actually inlines `ObjIsInstanceOfCached`, and makes a bit of a mess, because of the switch. I tried to massage it a little - replacing switch with ifs. That made codegen marginally better, but overall perf did not move, so I left the switch there.

I can try forceinlining `TryGet". Maybe it will change something... 

Overall, I think this kind of tuning should be done separately or we will never finish :-).

There are other opportunities - like some methods are now identical or nearly identical. Either use extra knowledge (like that we do array cast), or just fold the duplicates into one. 
Some helpers could do trivial checks (if there are such) before going directly to cache. That is assuming the checks would be cheap and typically work. And so on.

Reply:
I have tried forceinline `TryGet`. On the casting benchmarks that we have I get +/- 1% difference with very few outliers. Outliers are mostly in trivial scenarios not affected by caching - those are more lightweight casts and sometimes have spikes. 

The inlined version has more cases where it is slower, but I do not see any common pattern when it would help/hurt a particular group of scenarios. 
I think the diffs are just noise.

My guess is the perf is currently dominated by memory accesses and logic. Inlining does not have as much effect. 

I think we should revisit this later when we move more helpers from asm to C. Then method boundaries may start matter more and we might squeeze something from this.

Reply:
Tried a bit more selective approach with inlining by providing two versions of `ObjIsInstanceOfCached`. 
- one without identity check and with inlined TryGet - to be used from callers that have done identity checks already (mostly asm helpers) 
- one with identity checks, itself inlineable into the caller, but with TryGet not inlined - to be used in callers that do not do identity checks themselves (mostly C helpers)

The result is roughly the same. It is hard to see if there is an improvement next to the noise.